### DISCUSSION TITLE
Tower Defense - Technical Review

### DISCUSSION THREAD

**Discussion by argastle** [Freaky2164/TowerDefense#50] [General]
## Date:
22.05.2023 

## Participants: 
Lukas Weber (Technical Lead)
Johannes Methfessel (Technical Lead)
Tim Wäckerle (Moderator)
Paul Faller (Timekeeper)
Nico Argast (Notemaker)

## Goal: 
The purpose of this code review is to assess the RoundHandler and EnemySpawner components in order to align with best practices, minimize code instability, and ensure functionality.

## Components for Review:
- RoundHandler 
- EnemySpawner 

For these components, the following criteria have been considered:
- code quality
- scalability
- maintainability
- functionality

## Review methodology:
The following steps were followed during the review process:

1. Examination of best practices.
2. Discussion with the developer regarding their implementation.
3. Evaluation of the components' functionality.
4. Review of the code, highlighting both positive and negative aspects.
5. Definition of next steps.
6. Summary of lessons learned.

## Example Review:
RoundHandler: The RoundHandler component should handle rounds independently without direct connections to other components. It should, however, be aware of whether a round is currently being played to avoid simultaneous rounds. Other components should be able to easily access information about the active round.

![TechnicalReview](https://github.com/Freaky2164/TowerDefense/assets/97842228/b3da19fa-156e-446b-a211-5ecd2019a28c)

Upon reviewing the code, it was identified that certain improvements are needed. Firstly, adherence to C# naming conventions should be addressed, specifically the inclusion of underscores in the naming of private properties. On a positive note, the implementation of event handling is commendable, as it allows convenient access to information about the active round in other components without the RoundHandler needing to be aware of them. A potential area for improvement is the initialization of rounds within the RoundHandler class, which should be extracted into a separate configuration file for future edits within the Unity editor.

### Actions:
- Extract information about rounds into a separate configuration file.
- Refactor the code to adhere to C# naming conventions.
- Responsible for these actions: The developer who created the RoundHandler component.

### Lessons learned:
- Emphasize encapsulation of components from each other.
- Clarify the responsibilities and scope of each component.

**Comment by KevStr06** [comment]
Hey team TD, 
Great review! The RoundHandler and EnemySpawner components were assessed based on code quality, scalability, maintainability, and functionality. The review followed a comprehensive methodology, including examining best practices, discussing implementation, and evaluating component functionality.

The example review for RoundHandler provides valuable feedback, emphasizing the need for independent round handling and easy access to round information. Naming convention issues and the suggestion to extract round information into a separate configuration file are practical improvements. The positive aspect of event handling is noted for promoting loose coupling.

The defined actions, such as extracting round information and adhering to naming conventions, provide clear guidelines for the responsible developer. The lessons learned highlight encapsulation and clarifying component responsibilities.

Overall, the review demonstrates a thorough evaluation and sets a good example for component reviews. Well done!

Team cybuy

**Comment by Mohmad-Naser-alnakeshbandi** [comment]
Hello TD Team, 

Excellent evaluation! The RoundHandler and EnemySpawner components underwent a meticulous assessment encompassing code quality, scalability, maintainability, and functionality. The review process was comprehensive, incorporating the examination of best practices, in-depth discussions on implementation, and a careful evaluation of component functionality.

The review for RoundHandler offers invaluable insights, highlighting the significance of independent round handling and convenient access to round information. Practical recommendations, such as resolving naming convention issues and extracting round information into a separate configuration file, are proposed for improvement. The review also acknowledges the positive aspect of event handling, which promotes loose coupling within the codebase.

The specified actions, such as the extraction of round information and adherence to naming conventions, provide clear guidelines for conscientious developers. The lessons learned from the review underscore the importance of encapsulation and clearly defining component responsibilities.

Best regards, 
Ghalia Team
